crates/assists/src/handlers/reorder_impl.rs

```diff
@@ -0,0 +1,263 @@
+        return None;
+    }
+
+    let target = items.syntax().text_range();
+    acc.add(AssistId("reorder_impl", AssistKind::RefactorRewrite), "Sort methods", target, |edit| {
+        let mut rewriter = algo::SyntaxRewriter::default();
+        for (old, new) in methods.iter().zip(&sorted) {
+            rewriter.replace(old.syntax(), new.syntax());
+        }
+        edit.rewrite(rewriter);
+    })
+}
+
+fn compute_method_ranks(
+    path: &Option<ast::Path>,
+    ctx: &AssistContext,
+    methods: &[Fn],
+) -> Option<FxHashMap<String, usize>> {
+    if let Some(path) = path {
+        let td = trait_definition(path, &ctx.sema)?;
+
+        Some(
+            td.items(ctx.db())
+                .iter()
+                .flat_map(|i| match i {
+                    hir::AssocItem::Function(f) => Some(f),
+                    _ => None,
+                })
+                .enumerate()
+                .map(|(idx, func)| ((func.name(ctx.db()).to_string(), idx)))
+                .collect(),
+        )
+    } else {
+        Some(
+            methods
+                .iter()
+                .flat_map(|f| f.name().map(|n| n.to_string()))
+                .sorted()
+                .enumerate()
+                .map(|(idx, name)| (name, idx))
+                .collect(),
+        )
+    }
+}
+
+fn trait_definition(path: &ast::Path, sema: &Semantics<RootDatabase>) -> Option<Trait> {
+    match sema.resolve_path(path)? {
+        PathResolution::Def(ModuleDef::Trait(trait_)) => Some(trait_),
+        _ => None,
+    }
+}
+
+fn get_methods(items: &AssocItemList) -> Vec<Fn> {
+    items
+        .assoc_items()
+        .flat_map(|i| match i {
+            ast::AssocItem::Fn(f) => Some(f),
+            _ => None,
+        })
+        .filter(|f| f.name().is_some())
+        .collect()
+}
+
+// trait Foo<T> {}
+
+// struct Bar;
+
+// impl Foo<usize> for Bar {}
+
+#[cfg(test)]
+mod tests {
+    use crate::tests::{check_assist, check_assist_not_applicable};
+
+    use super::*;
+
+    #[test]
+    fn reorder_impl_trait_methods() {
+        check_assist(
+            reorder_impl,
+            r#"
+        trait Bar {
+            fn a() {}
+            fn c() {}
+            fn b() {}
+            fn d() {}
+        }
+
+        struct Foo {c: usize}
+        <|>impl Bar for Foo {
+            fn d() {}
+            fn b() {}
+            fn c() {}
+            fn a() {}
+        }
+        "#,
+            r#"
+        trait Bar {
+            fn a() {}
+            fn c() {}
+            fn b() {}
+            fn d() {}
+        }
+
+        struct Foo {c: usize}
+        impl Bar for Foo {
+            fn a() {}
+            fn c() {}
+            fn b() {}
+            fn d() {}
+        }
+        "#,
+        )
+    }
+
+    #[test]
+    fn not_applicable_if_trait_empty() {
+        check_assist_not_applicable(
+            reorder_impl,
+            r#"
+        trait Bar {};
+        struct Foo;
+        <|>impl Bar for Foo {}
+        "#,
```

Comment:
```suggestion
trait Bar {};
struct Foo;
<|>impl Bar for Foo {}
"#,
```
Ideally unindent the test fixtures, see https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#minimal-tests

Reply:
I indented these test fixtures since I was copying the style / flow of `reorder_fields.rs`. Should I also remove the test indentation in `reorder_fields.rs`?

Reply:
Ah ye some tests are still indented, I'd say leave them as is as to not make the diff of this PR too big and keep it in scope. Unindenting the tests in `reorder_fields.rs` can be done in a separate PR.

Reply:
Sound good